crates/core/tedge/src/cli/mqtt/cli.rs

```diff
@@ -44,43 +43,18 @@ pub enum TEdgeMqttCli {
 
 impl BuildCommand for TEdgeMqttCli {
     fn build_command(self, context: BuildContext) -> Result<Box<dyn Command>, crate::ConfigError> {
-        let port = context
-            .config_repository
-            .load()?
-            .query(MqttClientPortSetting)?;
-        let host = context
-            .config_repository
-            .load()?
-            .query(MqttClientHostSetting)?;
-        let ca_file = context
-            .config_repository
-            .load()?
-            .query(MqttClientCafileSetting)
-            .ok();
-        let ca_path = context
-            .config_repository
-            .load()?
-            .query(MqttClientCapathSetting)
-            .ok();
-        let client_cert = context
-            .config_repository
-            .load()?
-            .query(MqttClientAuthCertSetting);
-        let client_private_key = context
-            .config_repository
-            .load()?
-            .query(MqttClientAuthKeySetting);
+        let config = context.config_repository.load_new()?;
+        let auth_config = config.mqtt_client_auth_configs()?;
 
-        let client_auth_config = if client_cert.is_err() && client_private_key.is_err() {
-            None
-        } else {
-            let client_cert = client_cert?;
-            let client_private_key = client_private_key?;
-            Some(ClientAuthConfig {
-                cert_file: client_cert,
-                key_file: client_private_key,
-            })
-        };
+        let client_auth_config =
+            if auth_config.cert_file.is_none() && auth_config.key_file.is_none() {
+                None
+            } else {
+                Some(ClientAuthConfig {
+                    cert_file: auth_config.cert_file.clone().unwrap_or_default(),
+                    key_file: auth_config.key_file.unwrap_or_default(),
+                })
+            };
```

Comment:
If a function `mqtt_client_auth_configs` got created in order to put this "if either client key or cert are not set, then both are not set" logic there, then couldn't we also move this bit there? Because it seems to me like we need to do a fairly simple transformation but go in circles instead.

What we essentially have is: `(ca_dir, ca_file, cert_file, key_file)`, all of them are optional. Then the logic is:

1. If either `ca_dir` or `ca_file` is set, we should use TLS to connect. The code that we call later is aware of this convention, so we should just pass these fields as they are
2. If either `cert_file` or `key_file` is not set while the other is set, set both to `None` and don't use client authentication (or if we want to be very correct, either print a warning or return an error)

I think all this behaviour could be encompassed by a single function, and the above could even be its documentation. The output could be descibed by a struct like this:

```
MqttAuthConfig {
  ca_dir: Utf8PathBuf,
  ca_file: Utf8PathBuf,
  client: MqttAuthClientConfig {
    cert_file: Utf8PathBuf,
    key_file: Utf8PathBuf\
  }
}
```

and the function signature could be `(ca_dir, ca_file, cert_file, key_file) -> MqttAuthConfig` (in this signature, I don't mean this 4-tuple to be a literal tuple, but just these 4 properties we get from tedge_config)

Reply:
Refactored it below

```
#[derive(Debug, Clone, Default)]
pub struct MqttAuthConfig {
    pub ca_dir: Option<Utf8PathBuf>,
    pub ca_file: Option<Utf8PathBuf>,
    pub client: Option<MqttAuthClientConfig>,
}

#[derive(Debug, Clone, Default)]
pub struct MqttAuthClientConfig {
    pub cert_file: Utf8PathBuf,
    pub key_file: Utf8PathBuf,
}
```

But I did not get what function you are talking about here. Because the `tedge config` provides these certificates, I populated these structs inside the `mqtt_client_auth_config` function, after some `validations` for the `MqttAuthClientConfig`

Reply:
`mqtt_client_auth_config()` function as implemented, does exactly what i wanted, just reading properties from the config and putting them into this struct so the later code doesn't have to run this logic again.